Unable to compile ADC for esp32-c3

### The problem

Since esphome version 2023.7.0 I get this error when compiling the ADC sensor.
I have tried cleaning the build files but the problem persists.

```
INFO ESPHome 2023.7.0
INFO Reading configuration /config/adc-esp32c3.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing adc-esp32c3 (board: esp32-c3-devkitm-1; framework: arduino; platform: platformio/espressif32@5.3.0)
--------------------------------------------------------------------------------
Library Manager: Installing esphome/AsyncTCP-esphome @ 1.2.2
INFO Installing esphome/AsyncTCP-esphome @ 1.2.2
Unpacking  [####################################]  100%
Library Manager: AsyncTCP-esphome@1.2.2 has been installed!
INFO AsyncTCP-esphome@1.2.2 has been installed!
Library Manager: Installing esphome/ESPAsyncWebServer-esphome @ 2.1.0
INFO Installing esphome/ESPAsyncWebServer-esphome @ 2.1.0
Unpacking  [####################################]  100%
Library Manager: ESPAsyncWebServer-esphome@2.1.0 has been installed!
INFO ESPAsyncWebServer-esphome@2.1.0 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/noise-c @ 0.1.4
INFO Installing esphome/noise-c @ 0.1.4
Unpacking  [####################################]  100%
Library Manager: noise-c@0.1.4 has been installed!
INFO noise-c@0.1.4 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10018.1
INFO Installing esphome/libsodium @ 1.10018.1
Unpacking  [####################################]  100%
Library Manager: libsodium@1.10018.1 has been installed!
INFO libsodium@1.10018.1 has been installed!
HARDWARE: ESP32C3 160MHz, 320KB RAM, 4MB Flash
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch5
Dependency Graph
|-- AsyncTCP-esphome @ 1.2.2
|-- WiFi @ 2.0.0
|-- FS @ 2.0.0
|-- Update @ 2.0.0
|-- ESPAsyncWebServer-esphome @ 2.1.0
|-- DNSServer @ 2.0.0
|-- ESPmDNS @ 2.0.0
|-- noise-c @ 0.1.4
Compiling .pioenvs/adc-esp32c3/src/esphome/components/adc/adc_sensor.cpp.o
Compiling .pioenvs/adc-esp32c3/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/adc-esp32c3/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/adc-esp32c3/src/esphome/components/api/api_pb2.cpp.o
src/esphome/components/adc/adc_sensor.cpp: In member function 'virtual float esphome::adc::ADCSensor::sample()':
src/esphome/components/adc/adc_sensor.cpp:164:55: error: invalid conversion from 'int32_t*' {aka 'long int*'} to 'int*' [-fpermissive]
       adc2_get_raw(channel2_, ADC_WIDTH_MAX_SOC_BITS, &raw);
                                                       ^~~~
In file included from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/config/.esphome/platformio/packages/framework-arduinoespressif32/tools/sdk/esp32c3/include/driver/include/driver/adc.h:381:81: note:   initializing argument 3 of 'esp_err_t adc2_get_raw(adc2_channel_t, adc_bits_width_t, int*)'
 esp_err_t adc2_get_raw(adc2_channel_t channel, adc_bits_width_t width_bit, int *raw_out);
                                                                            ~~~~~^~~~~~~
src/esphome/components/adc/adc_sensor.cpp:196:53: error: invalid conversion from 'int32_t*' {aka 'long int*'} to 'int*' [-fpermissive]
     adc2_get_raw(channel2_, ADC_WIDTH_MAX_SOC_BITS, &raw11);
                                                     ^~~~~~
In file included from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/config/.esphome/platformio/packages/framework-arduinoespressif32/tools/sdk/esp32c3/include/driver/include/driver/adc.h:381:81: note:   initializing argument 3 of 'esp_err_t adc2_get_raw(adc2_channel_t, adc_bits_width_t, int*)'
 esp_err_t adc2_get_raw(adc2_channel_t channel, adc_bits_width_t width_bit, int *raw_out);
                                                                            ~~~~~^~~~~~~
src/esphome/components/adc/adc_sensor.cpp:199:55: error: invalid conversion from 'int32_t*' {aka 'long int*'} to 'int*' [-fpermissive]
       adc2_get_raw(channel2_, ADC_WIDTH_MAX_SOC_BITS, &raw6);
                                                       ^~~~~
In file included from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/config/.esphome/platformio/packages/framework-arduinoespressif32/tools/sdk/esp32c3/include/driver/include/driver/adc.h:381:81: note:   initializing argument 3 of 'esp_err_t adc2_get_raw(adc2_channel_t, adc_bits_width_t, int*)'
 esp_err_t adc2_get_raw(adc2_channel_t channel, adc_bits_width_t width_bit, int *raw_out);
                                                                            ~~~~~^~~~~~~
src/esphome/components/adc/adc_sensor.cpp:202:57: error: invalid conversion from 'int32_t*' {aka 'long int*'} to 'int*' [-fpermissive]
         adc2_get_raw(channel2_, ADC_WIDTH_MAX_SOC_BITS, &raw2);
                                                         ^~~~~
In file included from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/config/.esphome/platformio/packages/framework-arduinoespressif32/tools/sdk/esp32c3/include/driver/include/driver/adc.h:381:81: note:   initializing argument 3 of 'esp_err_t adc2_get_raw(adc2_channel_t, adc_bits_width_t, int*)'
 esp_err_t adc2_get_raw(adc2_channel_t channel, adc_bits_width_t width_bit, int *raw_out);
                                                                            ~~~~~^~~~~~~
src/esphome/components/adc/adc_sensor.cpp:205:59: error: invalid conversion from 'int32_t*' {aka 'long int*'} to 'int*' [-fpermissive]
           adc2_get_raw(channel2_, ADC_WIDTH_MAX_SOC_BITS, &raw0);
                                                           ^~~~~
In file included from src/esphome/components/adc/adc_sensor.h:10,
                 from src/esphome/components/adc/adc_sensor.cpp:1:
/config/.esphome/platformio/packages/framework-arduinoespressif32/tools/sdk/esp32c3/include/driver/include/driver/adc.h:381:81: note:   initializing argument 3 of 'esp_err_t adc2_get_raw(adc2_channel_t, adc_bits_width_t, int*)'
 esp_err_t adc2_get_raw(adc2_channel_t channel, adc_bits_width_t width_bit, int *raw_out);
                                                                            ~~~~~^~~~~~~
*** [.pioenvs/adc-esp32c3/src/esphome/components/adc/adc_sensor.cpp.o] Error 1
========================== [FAILED] Took 6.42 seconds ==========================
```


### Which version of ESPHome has the issue?

2023.7.0

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2023.7.2

### What platform are you using?

ESP32

### Board

ESP32C3 (Lolin C3 Mini)

### Component causing the issue

adc

### Example YAML snippet

```yaml
esphome:
  name: adc-esp32c3
  friendly_name: adc-esp32c3

esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: arduino

# Enable logging
logger:

# Enable Home Assistant API
api:
  encryption:
    key: "***************************************"

ota:
  password: "*************************************"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Adc-Esp32C3 Fallback Hotspot"
    password: "**********"

captive_portal:

sensor:
  - platform: adc
    pin: GPIO4
    name: "Voltaje módulo Lluvia"
    id: "lluvia_analogica"
    attenuation: auto
    unit_of_measurement: "V"
    update_interval: 60s
```


### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

--- a/esphome/components/adc/adc_sensor.cpp
+++ b/esphome/components/adc/adc_sensor.cpp
@@ -32,8 +32,8 @@ static const int32_t SOC_ADC_RTC_MAX_BITWIDTH = 12;
 #endif
 #endif
 
-static const int32_t ADC_MAX = (1 << SOC_ADC_RTC_MAX_BITWIDTH) - 1;    // 4095 (12 bit) or 8191 (13 bit)
-static const int32_t ADC_HALF = (1 << SOC_ADC_RTC_MAX_BITWIDTH) >> 1;  // 2048 (12 bit) or 4096 (13 bit)
+static const int ADC_MAX = (1 << SOC_ADC_RTC_MAX_BITWIDTH) - 1;    // 4095 (12 bit) or 8191 (13 bit)
+static const int ADC_HALF = (1 << SOC_ADC_RTC_MAX_BITWIDTH) >> 1;  // 2048 (12 bit) or 4096 (13 bit)
 #endif
 
 #ifdef USE_RP2040
@@ -59,7 +59,7 @@ extern "C"
   }
 
   // load characteristics for each attenuation
-  for (int32_t i = 0; i < (int32_t) ADC_ATTEN_MAX; i++) {
+  for (int32_t i = 0; i <= ADC_ATTEN_DB_11; i++) {
     auto adc_unit = channel1_ != ADC1_CHANNEL_MAX ? ADC_UNIT_1 : ADC_UNIT_2;
     auto cal_value = esp_adc_cal_characterize(adc_unit, (adc_atten_t) i, ADC_WIDTH_MAX_SOC_BITS,
                                               1100,  // default vref
@@ -157,7 +157,7 @@ float ADCSensor::sample() {
 #ifdef USE_ESP32
 float ADCSensor::sample() {
   if (!autorange_) {
-    int32_t raw = -1;
+    int raw = -1;
     if (channel1_ != ADC1_CHANNEL_MAX) {
       raw = adc1_get_raw(channel1_);
     } else if (channel2_ != ADC2_CHANNEL_MAX) {
@@ -174,7 +174,7 @@ float ADCSensor::sample() {
     return mv / 1000.0f;
   }
 
-  int32_t raw11 = ADC_MAX, raw6 = ADC_MAX, raw2 = ADC_MAX, raw0 = ADC_MAX;
+  int raw11 = ADC_MAX, raw6 = ADC_MAX, raw2 = ADC_MAX, raw0 = ADC_MAX;
 
   if (channel1_ != ADC1_CHANNEL_MAX) {
     adc1_config_channel_atten(channel1_, ADC_ATTEN_DB_11);

--- a/esphome/components/adc/adc_sensor.h
+++ b/esphome/components/adc/adc_sensor.h
@@ -62,7 +62,7 @@ class ADCSensor : public sensor::Sensor, public PollingComponent, public voltage
   adc1_channel_t channel1_{ADC1_CHANNEL_MAX};
   adc2_channel_t channel2_{ADC2_CHANNEL_MAX};
   bool autorange_{false};
-  esp_adc_cal_characteristics_t cal_characteristics_[(int32_t) ADC_ATTEN_MAX] = {};
+  esp_adc_cal_characteristics_t cal_characteristics_[ADC_ATTEN_MAX] = {};
 #endif
 };
 

--- a/tests/test7.yaml
+++ b/tests/test7.yaml
@@ -31,3 +31,11 @@ logger:
 http_request:
   useragent: esphome/tagreader
   timeout: 10s
+
+sensor:
+  - platform: adc
+    id: adc_sensor_p4
+    name: ADC pin 4
+    pin: 4
+    attenuation: 11db
+    update_interval: 1s

